From: Yury Norov <yury.norov@gmail.com>
Subject: lib/group_cpus: optimize inner loop in grp_spread_init_one()
Date: Thu, 7 Dec 2023 12:38:57 -0800

The loop starts from the beginning every time we switch to the next
sibling mask.  This is the Schlemiel the Painter's style of coding because
we know for sure that nmsk is clear up to current CPU, and we can just
continue from the next CPU.

Also, we can do it nicer if leverage the dedicated for_each() iterator.

Link: https://lkml.kernel.org/r/20231207203900.859776-4-yury.norov@gmail.com
Signed-off-by: Yury Norov <yury.norov@gmail.com>
Cc: Andy Shevchenko <andriy.shevchenko@linux.intel.com>
Cc: Ming Lei <ming.lei@redhat.com>
Cc: Rasmus Villemoes <linux@rasmusvillemoes.dk>
Cc: Thomas Gleixner <tglx@linutronix.de>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
---

 lib/group_cpus.c |   10 +++++-----
 1 file changed, 5 insertions(+), 5 deletions(-)

--- a/lib/group_cpus.c~lib-group_cpus-optimize-inner-loop-in-grp_spread_init_one
+++ a/lib/group_cpus.c
@@ -30,13 +30,13 @@ static void grp_spread_init_one(struct c
 
 		/* If the cpu has siblings, use them first */
 		siblmsk = topology_sibling_cpumask(cpu);
-		for (sibl = -1; cpus_per_grp > 0; ) {
-			sibl = cpumask_next(sibl, siblmsk);
-			if (sibl >= nr_cpu_ids)
-				break;
+		sibl = cpu + 1;
+
+		for_each_cpu_and_from(sibl, siblmsk, nmsk) {
 			__cpumask_clear_cpu(sibl, nmsk);
 			__cpumask_set_cpu(sibl, irqmsk);
-			cpus_per_grp--;
+			if (cpus_per_grp-- == 0)
+				return;
 		}
 	}
 }
_
